src/Symfony/Component/Cache/Adapter/TagAware/FilesystemTagAwareAdapter.php

```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * @var \Symfony\Component\Filesystem\Filesystem|null
+     */
+    private $fs;
+
+    public function __construct(string $namespace = '', int $defaultLifetime = 0, string $directory = null, MarshallerInterface $marshaller = null)
+    {
+        $this->marshaller = $marshaller ?? new DefaultMarshaller();
+        parent::__construct('', $defaultLifetime);
+        $this->init($namespace, $directory);
+    }
+
+    /**
+     * This method overrides {@see \Symfony\Component\Cache\Traits\FilesystemTrait::doSave}.
+     *
+     * It needs to be overridden due to:
+     * - usage of `serialize()` in the original method
+     * - need to store tag information on save
+     *
+     * {@inheritdoc}
+     */
+    protected function doSave(array $values, $lifetime)
+    {
+        $failed = $this->saveCache($values, $lifetime);
+
+        // Save Tags as symlinks, uses Filesystem Component to let it handle exceptions correctly
+        $fs = $this->getFilesystem();
+        foreach ($values as $id => $value) {
+            if (!empty($failed) && \in_array($id, $failed)) {
+                continue;
+            }
+
+            $file = $this->getFile($id);
+            foreach ($value['tags'] as $tag) {
+                $fs->symlink(
+                    $file,
+                    $this->getTagFolder($tag).$this->getTagIdFile($id));
+            }
+        }
+
+        return $failed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function invalidateTags(array $tags)
+    {
+        if (empty($tags)) {
+            return;
+        }
+
+        foreach (array_unique($tags) as $tag) {
+            $tagsFolder = $this->getTagFolder($tag);
+            if (!is_dir($tagsFolder)) {
+                continue;
+            }
+
+            foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->getTagFolder($tag), \FilesystemIterator::SKIP_DOTS)) as $itemLink) {
+                if (!$itemLink->isLink()) {
+                    throw new \Exception('Tag link is not a link: '.$itemLink);
+                }
+
+                $valueFile = $itemLink->getRealPath();
+                if ($valueFile && file_exists($valueFile)) {
+                    @unlink($valueFile);
+                }
+
+                @unlink((string) $itemLink);
+            }
+        }
+
+        return true;
+    }
+
+    private function getFilesystem(): Filesystem
+    {
+        return $this->fs ?? $this->fs = new Filesystem();
+    }
+
+    private function getTagFolder($tag): string
+    {
+        return $this->directory.self::TAG_FOLDER.\DIRECTORY_SEPARATOR.str_replace('/', '-', $tag).\DIRECTORY_SEPARATOR;
+    }
+
+    private function getTagIdFile($id): string
+    {
+        // Use MD5 to favor speed over security, which is not an issue here
+        $hash = str_replace('/', '-', base64_encode(hash('md5', static::class.$id, true)));
+
+        return substr($hash, 0, 20);
+    }
+
+    /**
+     * @internal for unit tests only
```

Comment:
Shouldn’t this be done via reflection in tests?

Reply:
I agree